libraries/amqp/include/eosio/amqp/transactional_amqp_publisher.hpp

```diff
@@ -0,0 +1,55 @@
+#include <fc/time.hpp>
+
+#include <string>
+#include <memory>
+#include <vector>
+#include <functional>
+#include <deque>
+
+namespace eosio {
+
+/*
+ * A transactional_amqp_publisher manages a connection to a single AMQP server that publishes messages to an exchange
+ * in a reliable manner: connections to the AMQP server are retried, publishing occurs in a transaction,
+ * publish blocks until AMQP acks transactional group of messages.
+ *
+ * on_fatal_error is called if AMQP does not ack transactional group of messages in the specified time. Called from
+ * AMQP thread or calling thread.
+ *
+ * Be aware that the transactional_amqp_publisher will NOT create the exchange. It requires the exchange
+ * to already exist. Publishing to a non-existing exchange will effectively block as the transaction will not be
+ * ack'ed.
+ */
+
+class transactional_amqp_publisher {
+   public:
+      // Called from amqp thread when publish messages are not ack
+      using error_callback_t = std::function<void(const std::string& err)>;
+
+      /// Create a reliable queue to the given server publishing to the given exchange
+      /// \param server_url server url as amqp://...
+      /// \param exchange the exchange to publish to
+      /// \param time_out time to wait on transactional ack from AMQP before calling on_fatal_error
+      /// \param on_fatal_error called from AMQP does not ack transaction in time_out time
+      transactional_amqp_publisher(const std::string& server_url, const std::string& exchange,
+                                   const fc::microseconds& time_out, error_callback_t on_fatal_error);
+
+      /// Publish messages. May be called from any thread except internal thread (do not call from on_fatal_error)
+      /// Blocks until AMQP ack of success transactional commit or on_fatal_error if timeout waiting on ack
```

Comment:
This implementation may split the messages into multiple transactions.  The previous implementation using `reliable_amqp_publisher` guaranteed that they were all in the same transaction.  Is that an acceptable change?

Reply:
Thanks for pointing that out. I meant to ask @spoonincode about that. Seems like we should just put them all into one transaction unless rabbitmq has something that makes putting into smaller needed.

Reply:
Changed to do all events of a block in one transaction.